cpp/src/arrow/io/memory.h

```diff
@@ -171,8 +171,8 @@ class ARROW_EXPORT BufferReader
 
   Result<int64_t> DoRead(int64_t nbytes, void* buffer);
   Result<std::shared_ptr<Buffer>> DoRead(int64_t nbytes);
-  Result<int64_t> DoReadAt(int64_t position, int64_t nbytes, void* out);
-  Result<std::shared_ptr<Buffer>> DoReadAt(int64_t position, int64_t nbytes);
+  virtual Result<int64_t> DoReadAt(int64_t position, int64_t nbytes, void* out);
+  virtual Result<std::shared_ptr<Buffer>> DoReadAt(int64_t position, int64_t nbytes);
```

Comment:
As said above, these two `DoReadAt` are changed to `virtual` so that they can be overridden and mock fs could use them for collecting the read stats.